src/Features/CSharp/Portable/ForeachToFor/CSharpForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,160 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.ForeachToFor;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.ForeachToFor
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpForeachToForCodeRefactoringProvider)), Shared]
+    internal sealed class CSharpForeachToForCodeRefactoringProvider : AbstractForeachToForCodeRefactoringProvider
+    {
+        public CSharpForeachToForCodeRefactoringProvider()
+        {
+        }
+
+        protected override SyntaxNode GetForeachStatement(SyntaxToken token)
+        {
+            var foreachStatement = token.Parent.FirstAncestorOrSelf<ForEachStatementSyntax>();
+            if (foreachStatement == null)
+            {
+                return null;
+            }
+
+            // support refactoring only if caret is in between "foreach" and ")"
+            var scope = TextSpan.FromBounds(foreachStatement.ForEachKeyword.Span.Start, foreachStatement.CloseParenToken.Span.End);
+            if (!scope.IntersectsWith(token.Span))
+            {
+                return null;
+            }
+
+            // check whether there is any comments between foreach and ) tokens
+            // if they do, we don't support conversion.
+            foreach (var trivia in foreachStatement.DescendantTrivia(n => n == foreachStatement || scope.Contains(n.FullSpan)))
```

Comment:
i don't think this is necessary.  we can support it.  it'll just lose some comments.  As this is a refactoring and not a fix, i don't think that's an issue.  If user's run into reasonable comment cases that are worth supporting, we can adapt the code to do the right thing if it's important enough.

Reply:
again, we can see how many people complaining about it and support it.

Reply:
This is not how we generally do IDE features.  We don't go and explicitly disable waiting for feedback.  We are permissive except for the issue that we know we can't address.  IN this case, we can support these cases, but you've gone and added code to explicitly disable the feature.  Very few features do this, and only for extremely important cases where the fix would actually be *very* bad.

Reply:
Why not support conversion? If really you want to preserve the comments, just put them anywhere, it's doesn't really matter that much. I don't see a problem here. For example:
```c#
foreach /*a*/ (/*b*/ var /*c*/ x /*d*/ in /*e*/ list /*f*/) /*g*/
````
could be converted to:
```c#
for /*a*/ (/*b*/ int i = 0; i < list.Count; i++ /*f*/) /*g*/
{
    var /*c*/ x /*d*/ = /*e*/ list[i];
}
```

If users don't like what we did, they can adjust the result themselves but we could still save half their work by doing the conversion. Once people know about this refactoring, **they'll expect it to show up everywhere and will blame the IDE for not working if it's not offered**.

Reply:
tracking issue here - https://github.com/dotnet/roslyn/issues/25667